### PR TITLE
[sc-4188][sc-4193][sc-4299] New methods

### PR DESCRIPTION
**Pull request by othree**
- Get payment/refund
- Cancel order
- Idempodency-Key

### CODE REVIEW

File: lib/smartpay/client.rb [hunk 1]
```diff
@@ -1,27 +1,54 @@
...
+                                                headers: headers.merge({Idempotency_Key: idempotency_key}).merge(params: request_params),
+                                                timeout: timeout,
+                                                payload: request_payload.to_json)
+        end
         JSON.parse(response.body, symbolize_names: true)
       end
 
       private
 
+      def nonce
```
Comment (carawangtw) [inline lib/smartpay/client.rb:48]: Minor suggestion: maybe we can name the method as `idempotency_key` and use it on lines 25 & 39 directly.

File: lib/smartpay/client.rb [hunk 2]
```diff
@@ -1,27 +1,54 @@
...
 
-      def post(path, payload = {})
+      def post(path, params: {}, payload: {})
+        request_params = default_params.merge(params)
+        request_payload = default_payload.merge(payload)
+        idempotency_key = nonce
+        response = with_retries(:max_tries => 3, :rescue => [RestClient::BadRequest, RestClient::BadGateway, RestClient::ServiceUnavailable, RestClient::GatewayTimeout]) do
+          RestClient::Request.execute(method: :post, url: api_url(path),
+                                                params: request_params,
+                                                headers: headers.merge({Idempotency_Key: idempotency_key}).merge(params: request_params),
```
Comment (carawangtw) [inline lib/smartpay/client.rb:25]: Not sure if we need to pass the request params in POST and PUT requests.

FYI. The `params` key in this field will be the query string in the HTTP request.

ref: https://github.com/rest-client/rest-client#passing-advanced-options

```Due to unfortunate choices in the original API, the params used to populate the query string are actually taken out of the headers hash. So if you want to pass both the params hash and more complex options, use the special key :params in the headers hash. This design may change in a future major release.

RestClient::Request.execute(method: :get, url: 'http://example.com/resource',
                            timeout: 10, headers: {params: {foo: 'bar'}})

➔ GET http://example.com/resource?foo=bar
```

File: lib/smartpay/client.rb [hunk 3]
```diff
@@ -1,27 +1,54 @@
...
       end
 
-      def post(path, payload = {})
+      def post(path, params: {}, payload: {})
+        request_params = default_params.merge(params)
+        request_payload = default_payload.merge(payload)
+        idempotency_key = nonce
+        response = with_retries(:max_tries => 3, :rescue => [RestClient::BadRequest, RestClient::BadGateway, RestClient::ServiceUnavailable, RestClient::GatewayTimeout]) do
+          RestClient::Request.execute(method: :post, url: api_url(path),
+                                                params: request_params,
```
Comment (carawangtw) [inline lib/smartpay/client.rb:24]: Q: Is this argument work for the method?

File: lib/smartpay/client.rb [hunk 4]
```diff
@@ -1,27 +1,54 @@
...
+      def put(path, params: {}, payload: {})
+        request_params = default_params.merge(params).merge({'Idempotency-Key': nonce})
         request_payload = default_payload.merge(payload)
-        response = RestClient::Request.execute(method: :post, url: api_url(path), headers: headers, timeout: timeout,
-                                               payload: request_payload.to_json)
+        idempotency_key = nonce
+        response = with_retries(:max_tries => 3, :rescue => [RestClient::InternalServerError, RestClient::BadGateway, RestClient::ServiceUnavailable, RestClient::GatewayTimeout]) do
+          RestClient::Request.execute(method: :put, url: api_url(path),
+                                                params: request_params,
+                                                headers: headers.merge({Idempotency_Key: idempotency_key}).merge(params: request_params),
```
Comment (carawangtw) [inline lib/smartpay/client.rb:39]: ditto

File: spec/integration/api_spec.rb [hunk 5]
```diff
@@ -61,9 +61,12 @@
           successUrl: 'https://docs.smartpay.co/example-pages/checkout-successful',
           cancelUrl: 'https://docs.smartpay.co/example-pages/checkout-canceled',
         })
+
         expect(session.response).not_to be_empty
         expect(session.redirect_url).not_to be_empty
 
+        puts session.response
```
Comment (carawangtw) [inline spec/integration/api_spec.rb:68]: Is this for debug messages? Should we remove this line?

**Review by carawangtw — COMMENTED** [review]
Can we add unit tests, too? Thanks!

File: lib/smartpay/client.rb (hunk 2 above)
Comment (othree) [reply lib/smartpay/client.rb:25]: Yes, we need it for dev-lang and sdk-version

File: spec/integration/api_spec.rb (hunk 5 above)
Comment (othree) [reply spec/integration/api_spec.rb:68]: This is intended to show the checkout session to the console.

File: lib/smartpay/client.rb (hunk 3 above)
Comment (othree) [reply lib/smartpay/client.rb:24]: Which argument? Idempotency_Key? Yes, it is for POST/PUT/DELETE.

**Pull request merged**